File: src/types/runtime.ts
export type AssetType = 'VM' | 'OCP_POD' | 'MQ_QMGR' | 'MONGO_NODE' | 'ORACLE_DB' | 'KAFKA_BROKER' | 'SERVER';
export type TechStack = 'ibm_mq' | 'mongodb' | 'oracle' | 'mssql' | 'kafka' | 'vm' | 'ocp';
export type AssetEnvironment = 'UAT' | 'PRODUCTION' | 'DR';
export type OperationalState = 'ACTIVE' | 'STANDBY' | 'UNKNOWN';
export type ReplicationRole = 'PRIMARY' | 'SECONDARY' | 'PHYSICAL_STANDBY' | 'NONE' | 'MONGOS' | 'CONFIG_SVR' | 'SHARD_PRIMARY' | 'SHARD_SECONDARY';
export type ConfidenceLevel = 1 | 2 | 3 | 4;
export type FreshnessStatus = 'FRESH' | 'STALE' | 'VERY_STALE' | 'UNKNOWN';
export type DataSourceName = 'ibm_mq' | 'mongodb' | 'oracle_oem' | 'cmdb' | 'mssql' | 'kafka' | 'avi_loadbalancer' | 'scom' | 'ocp' | 'batch' | 'appdynamics';
export type ImportStatus = 'SUCCESS' | 'PARTIAL' | 'FAILED';

export interface RuntimeDataCenter {
  id: string;
  name: string;
  short_name?: string;
  region?: string;
  zone?: string;
  asset_count: number;
}

export interface RuntimeAsset {
  id: string;
  name: string;
  asset_type: AssetType;
  tech_stack: TechStack;
  environment: AssetEnvironment;
  host?: string;
  port?: number;
  platform?: string;
  data_center?: RuntimeDataCenter;
  latest_confidence_level?: ConfidenceLevel;
  latest_operational_state?: OperationalState;
  latest_replication_role?: ReplicationRole;
  write_authority?: boolean;
  last_seen_at?: string;
  is_deterministic?: boolean;
  data_source?: DataSourceName;
  metadata?: Record<string, string>;
}

export interface ApplicationComponent {
  id: string;
  application_id: string;
  application_name: string;
  component_name: string;
  component_type: 'DATABASE' | 'MESSAGING' | 'COMPUTE' | 'STORAGE';
  tech_stack: TechStack;
  assets: RuntimeAsset[];
}

export type AlignmentStatus = 'ALIGNED' | 'DRIFTED' | 'UNKNOWN';
export type ConfidenceLabelStr = 'HIGH' | 'MEDIUM' | 'LOW' | 'CONFLICT' | 'UNKNOWN';

export interface ApplicationLocationSummary {
  application_id: string;
  application_name: string;
  environment: AssetEnvironment;
  data_centers: string[];
  primary_write_dc?: string;
  overall_confidence: ConfidenceLevel;
  confidence_label?: ConfidenceLabelStr;
  confidence_score?: number;
  component_count: number;
  asset_count: number;
  stale_source_count: number;
  missing_source_count?: number;
  last_updated?: string;
  tech_stacks?: TechStack[];
  alignment_status?: AlignmentStatus;
}

export interface DataSourceInfo {
  source_name: DataSourceName;
  display_name: string;
  status: FreshnessStatus;
  record_count: number;
  last_import?: string;
  topology_confidence: ConfidenceLevel;
  traffic_confidence: ConfidenceLevel;
}

export interface SourceConflict {
  asset_name: string;
  source_a: { name: string; says: string };
  source_b: { name: string; says: string };
  last_checked: string;
}

export interface ApplicationLocationDetail {
  application_id: string;
  application_name: string;
  environment: AssetEnvironment;
  overall_confidence: ConfidenceLevel;
  confidence_label?: ConfidenceLabelStr;
  confidence_score?: number;
  components: ApplicationComponent[];
  data_sources: DataSourceInfo[];
  conflicts: SourceConflict[];
}

export interface RuntimeSnapshot {
  id: string;
  asset_id: string;
  snapshot_time: string;
  operational_state: OperationalState;
  replication_role?: ReplicationRole;
  data_source: DataSourceName;
  confidence_level: ConfidenceLevel;
  is_deterministic: boolean;
}

export interface DataSourceImport {
  id: string;
  source_name: DataSourceName;
  file_name: string;
  imported_at: string;
  record_count: number;
  status: ImportStatus;
  errors: string[];
}

export type ReplicationModel = 'SINGLE_WRITER' | 'MULTI_WRITER' | 'READ_REPLICA' | 'EVENTUAL' | 'UNKNOWN';
export type FailoverType = 'AUTOMATIC' | 'MANUAL' | 'NONE';
export type DriftType = 'MISSING_DC' | 'WRONG_PRIMARY' | 'MISSING_COMPONENT' | 'EXTRA_DC' | 'ROLE_MISMATCH' | 'STALE_DATA';
export type DriftSeverity = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export interface ApplicationIntent {
  application_id: string;
  application_name: string;
  intended_active_dcs: string[];
  intended_primary_dc: string;
  intended_environments: AssetEnvironment[];
  failover_type: FailoverType;
  replication_model: ReplicationModel;
  required_tech_stacks: TechStack[];
  alignment_status?: AlignmentStatus;
  owner?: string;
  notes?: string;
  last_review_date?: string;
  created_at: string;
  updated_at: string;
}

export interface IntentDrift {
  id: string;
  application_id: string;
  environment: AssetEnvironment;
  drift_type: DriftType;
  description: string;
  severity: DriftSeverity;
  intended: string;
  actual: string;
  detected_at: string;
}

export type AuditEventType =
  | 'IMPORT'
  | 'STATE_CHANGE'
  | 'CONFLICT_DETECTED'
  | 'INTENT_CREATED'
  | 'INTENT_UPDATED'
  | 'DRIFT_DETECTED'
  | 'PROPOSAL_SUBMITTED'
  | 'SEED_LOADED';

export interface AuditLogEntry {
  id: string;
  event_type: AuditEventType;
  application_id?: string;
  asset_name?: string;
  description: string;
  actor: string;
  source?: string;
  before_value?: string;
  after_value?: string;
  occurred_at: string;
}

export type ProposalStatus = 'PENDING' | 'ACCEPTED' | 'REJECTED';

export interface SourceProposal {
  id: string;
  source_name: string;
  system: string;
  signal_type: string;
  tech_stack: string;
  rationale: string;
  is_deterministic_claim: boolean;
  proposed_by: string;
  proposed_at: string;
  status: ProposalStatus;
}

File: backend/app/models/runtime.py
from sqlalchemy import Column, String, Integer, Boolean, ForeignKey, DateTime, Text, JSON, Float
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid
from app.db.base import Base

class RuntimeDataCenter(Base):
    __tablename__ = "runtime_data_centers"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String, nullable=False)
    short_name = Column(String, unique=True, index=True, nullable=False)
    region = Column(String, nullable=True)
    zone = Column(String, nullable=True)
    asset_count = Column(Integer, default=0)

    assets = relationship("RuntimeAsset", back_populates="data_center", cascade="all, delete-orphan")


class RuntimeAsset(Base):
    __tablename__ = "runtime_assets"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String, index=True, nullable=False)
    asset_type = Column(String, nullable=False)  # e.g., MQ_QMGR, MONGO_NODE, ORACLE_DB, KAFKA_BROKER, OCP_POD, SERVER
    tech_stack = Column(String, nullable=False)  # e.g., ibm_mq, mongodb, oracle, mssql, kafka, ocp, vm
    environment = Column(String, nullable=False)  # e.g., PRODUCTION, UAT, DR
    host = Column(String, nullable=True)
    port = Column(Integer, nullable=True)
    platform = Column(String, default="LINUX")
    
    data_center_short = Column(String, ForeignKey("runtime_data_centers.short_name", ondelete="CASCADE"), nullable=True)
    
    latest_confidence_level = Column(Integer, default=3)  # 1 - 4
    confidence_label = Column(String, default="MEDIUM")  # HIGH | MEDIUM | LOW | CONFLICT | UNKNOWN
    confidence_score = Column(Integer, default=65)        # 0-100 engine-computed score
    latest_operational_state = Column(String, default="ACTIVE")  # ACTIVE, STANDBY, UNKNOWN
    latest_replication_role = Column(String, default="NONE")  # PRIMARY, SECONDARY, PHYSICAL_STANDBY, MONGOS, etc.
    write_authority = Column(Boolean, default=False)
    last_seen_at = Column(DateTime, default=datetime.utcnow)
    is_deterministic = Column(Boolean, default=True)
    data_source = Column(String, nullable=False)  # e.g., cmdb, mongodb, oracle_oem, ibm_mq
    metadata_json = Column(JSON, nullable=True)  # Store specific columns like cluster, rs_nm, node chains

    data_center = relationship("RuntimeDataCenter", back_populates="assets")


class DataSourceImport(Base):
    __tablename__ = "data_source_imports"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    source_name = Column(String, nullable=False)  # e.g. ibm_mq, mongodb, oracle_oem, cmdb
    file_name = Column(String, nullable=False)
    imported_at = Column(DateTime, default=datetime.utcnow)
    record_count = Column(Integer, default=0)
    status = Column(String, default="SUCCESS")  # SUCCESS, PARTIAL, FAILED
    errors = Column(JSON, nullable=True)


class ApplicationIntent(Base):
    __tablename__ = "application_intents"

    application_id = Column(String, primary_key=True, index=True)
    application_name = Column(String, nullable=False)
    intended_active_dcs = Column(JSON, nullable=False)  # e.g., ["IBB1", "SHV"]
    intended_primary_dc = Column(String, nullable=False)  # e.g., "IBB1"
    intended_environments = Column(JSON, nullable=True)  # e.g., ["PRODUCTION"]
    failover_type = Column(String, default="AUTOMATIC")
    replication_model = Column(String, default="SINGLE_WRITER")  # SINGLE_WRITER, MULTI_WRITER
    required_tech_stacks = Column(JSON, nullable=False)  # e.g., ["oracle", "ibm_mq"]
    alignment_status = Column(String, default="UNKNOWN")  # ALIGNED | DRIFTED | UNKNOWN
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)


class SourceProposal(Base):
    __tablename__ = "source_proposals"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    source_name = Column(String, nullable=False)
    system = Column(String, nullable=False)
    signal_type = Column(String, nullable=False)
    tech_stack = Column(String, nullable=False)
    rationale = Column(Text, nullable=False)
    is_deterministic_claim = Column(Boolean, default=True)
    proposed_by = Column(String, nullable=False)
    proposed_at = Column(DateTime, default=datetime.utcnow)
    status = Column(String, default="PENDING")  # PENDING, ACCEPTED, REJECTED


class RuntimeAuditLog(Base):
    __tablename__ = "runtime_audit_logs"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    event_type = Column(String, nullable=False)  # IMPORT, DRIFT_DETECTED, CONFLICT_DETECTED, INTENT_CREATED, etc.
    description = Column(Text, nullable=False)
    actor = Column(String, default="operator")
    source = Column(String, nullable=True)
    application_id = Column(String, nullable=True)
    occurred_at = Column(DateTime, default=datetime.utcnow)
